### ISSUE TITLE
Does dopamine set a random seed?

### ISSUE THREAD

**Issue by halleanwoo** [google/dopamine#96]
Does dopamine set a random seed?

**Comment by psc-g** [comment]
we don't set one explicitly. is that what you are referring to?

**Comment by halleanwoo** [comment]
Thanks for your replay. If I want to compare results of two ideas, should I set a random seed?

**Comment by psc-g** [comment]
if you want to compare with the same seed, then yes. otherwise tensorflow will use a different seed each time.
but in general it's preferable to do a few independent runs and report mean + confidence intervals

**Comment by halleanwoo** [comment]
I see，thanks for your  patience.
May a nice day

**Comment by RylanSchaeffer** [comment]
@psc-g , what's the proper way to set a seed within Dopamine? The pretrained JAX agents (https://github.com/google/dopamine/tree/master/dopamine/jax) seem to be 5 different runs, which I assume corresponds to 5 seeds?

**Comment by psc-g** [comment]
hi rylan,
you can change the seeds for JAX agents by modifying the parameter to
PRNGKey here:
https://github.com/google/dopamine/blob/master/dopamine/jax/agents/dqn/dqn_agent.py#L287

unfortunately, this is not the only source of non-determinism. to name two
important source:
- there can be non-determinism when training with GPUs (e.g. this article
<https://www.twosigma.com/articles/a-workaround-for-non-determinism-in-tensorflow/>
)
- if you're running with sticky actions, there is a source of randomness
there
<https://github.com/openai/atari-py/blob/5ca24caf395320ead8c77440fe729cf53804616a/atari_py/ale_interface/src/environment/stella_environment.cpp#L155>
and there are likely others i'm missing.

hope this helps!

**Comment by RylanSchaeffer** [comment]
Hi Pablo, sorry I should have clarified - is there a way to change the seeds for TF agents? I guess if there's always stochasticity, this matters much less than I thought

**Comment by psc-g** [comment]
you could try this?
https://www.tensorflow.org/api_docs/python/tf/compat/v1/set_random_seed
caveat: i've never used that function.
